File: sliding_window.js
const arr = [5, 2, -1, 0, 3];
const k = 3;
const n = arr.length;

function maxSum1(arr, n, k) {
	let maxSum = Number.MIN_SAFE_INTEGER;

	for (let i = 0; i < n - k + 1; i++) {
		let currentSum = 0;
		for (let j = 0; j < k; j++) {
			currentSum += arr[i + j];
		}
		maxSum = Math.max(currentSum, maxSum);
	}

	return maxSum;
}

function maxSum(arr, n, k) {
	if (n < k) {
		console.log("invalid");
		return -1;
	}

	let windowSum = 0;
	for (let i = 0; i < k; i ++) {
		windowSum += arr[i];
	}

	let maxSum = windowSum;

	for (let j = k; j < n; j++) {
		windowSum += arr[j] - arr[j - k]
		maxSum = Math.max(maxSum, windowSum)
	}

	return maxSum;
}
// console.log(maxSum1(arr, n, k))
console.log(maxSum(arr, n, k))

File: sliding_window.rb
arr = [5, 2, -1, 0, 3]
k = 2
def max_sum(arr, k)
  n = arr.length
  if (k > n)
    puts "Invalid!"
    return
  end

  window_sum = 0
  (0...k).each do |index|
    window_sum += arr[index]
  end

  max_sum = window_sum

  (k...n).each do |i|
    window_sum = arr[i] - arr[i - k]
    max_sum = [max_sum, window_sum].max
  end

  return max_sum
end

puts max_sum(arr, k)
